Created package for vertex cover algorithms

jkinable opened the pull request:
In anticipation of pull request #227 I've created a new package for Vertex Cover algorithms, analogous to the Max Flow algorithm package. Let me know what you think.
I'll make VertexCovers.java deprecated, and I'll update the test classes.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/VertexCoverAlgorithm.java:
```diff
@@ -0,0 +1,81 @@
+     * @param vertexWeightMap map containing non-negative weights for each vertex
+     * @return a vertex cover
+     */
+    VertexCover<V> getVertexCover(Map<V, Integer> vertexWeightMap);
```
Would it make sense to split this out to a subinterface (similar to MatchingAlgorithm/WeightedMatchingAlgorithm)?  That avoids the murkiness of implementations having to throw UnsupportedOperationException.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/vertexCover/GreedyVertexCover.java:
```diff
@@ -0,0 +1,119 @@
+ * 28-Jul-2016 : Moved to dedicated package (JK)
+ *
+ */
+package org.jgrapht.alg.vertexCover;
```
Let's avoid camelCase in package names.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/VertexCoverAlgorithm.java:
```diff
@@ -0,0 +1,81 @@
+    VertexCover<V> getVertexCover(Map<V, Integer> vertexWeightMap);
+
+
+    class VertexCover<V>{
```
It's a little confusing that both this result class and the concrete algorithm implementations end in "Cover".

jkinable commented:
- Revised package for vertex cover algorithms
  - Two new interfaces: MinimumVertexCoverAlgorithm and MinimumWeightedVertexCoverAlgorithm
  - Test suite has been updated accordingly
- Added greedy algorithm for the minimum weighted vertex cover problem
- Added Clarkson's 2-approximation algorithm for the minimum weighted vertex cover problem
- Added Bar-Yehuda and Even's 2-approximation algorithm for the minimum weighted vertex cover problem


Notes:
- Since MinimumVertexCoverAlgorithm is a special case of MinimumWeightedVertexCoverAlgorithm, MinimumWeightedVertexCoverAlgorithm extends MinimumVertexCoverAlgorithm.
-To avoid confusion: A MinimumVertexCoverAlgorithm returns a VertexCover; Any class that implements a MinimumVertexCoverAlgorithm ends with *VCImpl. @jsichi : I hope that this takes away the confusion. Feel free to propose any other naming scheme if you don't like this one :). Also, note that in case of Max Flow algorithms, we have a similar issue: an MaximumFlowAlgorithm returns a MaximumFlow, and is implemented by *MaximumFlow classes, e.g. PushRelabelMaximumFlow.
-For consistency, do we want: MinimumVertexCoverAlgorithm or VertexCoverAlgorithm? I personally prefer the first, but this is already getting inconsistent in the interfaces, e.g. MatchingAlgorithm (vs MaximumMatchingAlgorithm) and MaximumFlowAlgorithm (vs FlowAlgorithm)

jsichi commented:
The VCImpl naming scheme is fine.  My bad on not noticing the MaximumFlow confusion when I reviewed that.  Instead, as a separate issue:  at that time I was a stickler for requiring an interface (with a companion class MaximumFlowImpl).  I suggest that pattern because it makes it possible to proxy the entire interface.  Let me know what you think about that; it's not a big deal.

I'm fine with MinimumVertexCoverAlgorithm.  I don't think we're going to get perfect accuracy+consistency here; approximate algorithms don't even return an actual minimum.  So as long as the interface expresses the intention correctly (e.g. MatchingAlgorithm doesn't actually imply any intention about maximum), that seems fine to me.

jkinable commented:
This commit got a bit bigger than anticipated. I realized that I made an error in the implementation of both the Clarkson 2-approximation and the greedy algorithm: my implementations had a runtime complexity of O(|E|*|V|) instead of O(|E|*log|V|) so I had to redo them both. Turned out a little less trivial than I anticipated. Please do a thorough revision.

Updated change log:

- Revised package for vertex cover algorithms
  - Two new interfaces: MinimumVertexCoverAlgorithm and MinimumWeightedVertexCoverAlgorithm
  - Test suite has been updated accordingly
- Added greedy algorithm for the minimum weighted vertex cover problem
- Added Clarkson's 2-approximation algorithm for the minimum weighted vertex cover problem
- Added Bar-Yehuda and Even's 2-approximation algorithm for the minimum weighted vertex cover problem
- Replaced greedy algorithm for the Unweighted Vertex Cover problem as implemented by Linda Buisman (O(|E|*|V|)) by a faster (O(|E|*log|V|)) implementation
- Revised VertexDegreeComparator
  - Added order enum, simplified comparator
  - Added Unit test


Issues:
1. For most graph types, clone() creates a shallow copy of the graph (see clone() in AbstractBaseGraph). Problem: Interfaces such as UndirectedGraph (which extends the main interface Graph) do no support the clone() method. Furthermore, Object.clone() is protected, so you cannot invoke clone() on a UndirectedGraph
Proposal: add clone() to the main Graph interface (should be done in separate pull req.)


2. Wrt to the naming of the MaximumFlow algorithms. These algorithms have not yet been released in any JGraphT distribution; they only exist in the master branch. If you want I can rename the 2 implementation classes in a similar way as the vertex cover classes thereby giving them the extension: MFImpl
This will keep everything nicely consistent. (separate pull req)

3. If you want I can add an interface VertexCover and an implementation VertexCoverImpl similar to the interface MaximumFlow and its implementation MaximumFlowImpl. Personally I think it's a bit overkill and syntactically looks a bit odd (e.g. "new VertexCoverImpl()" versus nice and clean "new VertexCover()") (I know, this is rather subjective to taste... ). Nevertheless, I don't have strong feelings regarding this, so if you prefer the Interface version then I'll happily add the Interface and corresponding implementation. After all, it's only a minute work :)

4. It seems that a number of classes use a SubGraph to create a working copy of the graph. Although in some cases this is fine, in a number of cases this is a bad idea. The SubGraph is implemented to guarantee that both its vertices and edges are a subset of the edges in the original (base) graph and this adds a certain amount of overhead. For example, querying the degree of a vertex in an undirected graph (as was done repeatedly in the old implementation of the greedy vertex cover algorithm), takes O(degreeOf(v)) time, as opposed to O(1) in the base graph, rendering the implementation significantly slower for dense graphs. We should probably go over the algorithm inplementations to check where a SubGraph is used; in addition, some optimizations of SubGraph can be performed.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/VertexCovers.java:
```diff
@@ -57,6 +57,7 @@
  * @author Linda Buisman
  * @since Nov 6, 2003
  */
+@Deprecated
```
When adding the @Deprecated annotation, it's good to add a companion Javadoc @deprecated which directs developers to the replacement.

jsichi commented:
1.  We don't want to add cloning at the Graph interface level, because algorithms shouldn't assume that every graph implementation is clonable.  For example, if you have an adapter that lets you access a neo4j graph via JGraphT, the clone operator may not make sense (particularly for a read-only algorithm).  If working on a copy makes the most sense, create a new empty graph, and then use Graphs.addGraph.

2.  Yes, let's go with MFImpl.

3.  Thanks, let's go with the interface.  My main motivation is that since we're going to the effort to build out an interfaces package, we should undertake it as an exercise in API design (e.g. that's why java.util.Map.Entry is an interface rather than a class).  I'm sure we won't get it perfect, but best effort is worth it.

4.  Optimization of Subgraph would probably get us the most payoff in the long run, perhaps by adding caching (with a listener to automatically invalidate the cache?).

jkinable commented:
Updated change log:

- Revised package for vertex cover algorithms
  - Two new interfaces: MinimumVertexCoverAlgorithm and MinimumWeightedVertexCoverAlgorithm
  - Test suite has been updated accordingly
- Added greedy algorithm for the minimum weighted vertex cover problem
- Added Clarkson's 2-approximation algorithm for the minimum weighted vertex cover problem
- Added Bar-Yehuda and Even's 2-approximation algorithm for the minimum weighted vertex cover problem
- Replaced greedy algorithm for the Unweighted Vertex Cover problem as implemented by Linda Buisman (O(|E|*|V|)) by a faster (O(|E|*log|V|)) implementation
- Revised VertexDegreeComparator
  - Added order enum, simplified comparator
  - Added Unit test
- Added exact algorithm for Weighted Vertex Covers. (With contributions from Nils Nolberg (@nolberg), see pull request #227)

- Interface for Vertex Cover has been added
- Wrt cloning: - This is a bit problematic if the algorithm works on for example an UndirectedGraph. Copying an UndirectedGraph through Graphs.addGraph is impossible since the underlying graph type is hidden. E.g. both a pseudograph and a simplegraph can be UndirectedGraph's. Currently, the only option to create a shallow working copy is to use the SubGraph classes.
- good suggestion to optimize the Subgraph classes.

jsichi commented:
Copying:  the algorithm doesn't care about the underlying implementation of the source graph.  So it can choose whatever is appropriate for the target type.  For example, if the algorithm has no constraints, then it should just instantiate the least constrained type (e.g. Pseudograph) and copy to that.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/vertexcover/BarYehudaEvenTwoApproxVCImpl.java:
```diff
@@ -0,0 +1,102 @@
+        Set<V> cover=new LinkedHashSet<>();
+        double weight=0;
+        UndirectedGraph<V,E> copy= new UndirectedSubgraph<>(graph, null, null);
+//        UndirectedGraph<V,E> copy=(UndirectedGraph<V,E>)graph.clone();
```
Delete commented-out code.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/vertexcover/ClarksonTwoApproxVCImpl.java:
```diff
@@ -0,0 +1,117 @@
+            ux.addNeighbor(vx);
+            vx.addNeighbor(ux);
+
+            assert(ux.neighbors.get(vx) == vx.neighbors.get(ux)): " in an undirected graph, if vx is a neighbor of ux, then ux must be a neighbor of vx";
```
is it safe to use == here?

jkinable replied:
Currently it is since it is not possible to change the working graph outside this class, and all vertices are guaranteed to be unique. Nevertheless I'll change it to equals, just in case.

jsichi commented on jgrapht-core/src/test/java/org/jgrapht/alg/VertexCoverTest.java:
```diff
@@ -0,0 +1,264 @@
+     */
+    protected Graph<Integer, DefaultEdge> createRandomPseudoGraph()
+    {
+        // TODO: move random graph generator to be under GraphGenerator
```
There's already one there.

jkinable replied:
This is a remnant left by the previous author. I'll see whether I can swap it out for the one in the GraphGenerator package.

jkinable replied:
done

jsichi commented:
Let's do this!
